src/control/nav2_controller_dwb/src/dwb_controller.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_controller_dwb/dwb_controller.hpp"
+#include <string>
+#include <chrono>
+#include <memory>
+#include "dwb_core/exceptions.h"
+#include "nav_2d_utils/conversions.h"
+
+using namespace std::chrono_literals;
+using std::shared_ptr;
+using nav2_tasks::TaskStatus;
+using dwb_core::DWBLocalPlanner;
+using dwb_core::CostmapROSPtr;
+
+namespace nav2_controller_dwb
+{
+
+DwbController::DwbController()
+: nav2_tasks::FollowPathTaskServer("FollowPathNode")
+{
+}
+
+DwbController::~DwbController()
+{
+}
+
+TaskStatus
+DwbController::execute(const nav2_tasks::FollowPathCommand::SharedPtr /*command*/)
+{
+  try {
+    nav_2d_msgs::msg::Path2D path;
+    // get path from command
+    tf2_ros::Buffer tfBuffer;
+    tf2_ros::TransformListener tfListener(tfBuffer);
+    cm_ = std::make_shared<costmap_2d::Costmap2DROS>("local_costmap", tfBuffer);
+    auto nh = shared_from_this();
+    odom_sub_ = std::make_shared<nav_2d_utils::OdomSubscriber>(*this);
+    vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 1);
+    planner_.initialize(nh, shared_ptr<tf2_ros::Buffer>(&tfBuffer), cm_);
+    planner_.setPlan(path);
+    while (true) {
+      nav_2d_msgs::msg::Pose2DStamped pose2d;
+      if (!getRobotPose(pose2d)) {
+        publishZeroVelocity();
+      } else {
+        if (isGoalReached(pose2d)) {
+          break;
+        }
+        auto velocity = odom_sub_->getTwist();
+        auto cmd_vel_2d = planner_.computeVelocityCommands(pose2d, velocity);
+        publishVelocity(cmd_vel_2d);
+
+        if (cancelRequested()) {
+          RCLCPP_INFO(this->get_logger(), "execute: task has been canceled");
+          setCanceled();
+          return TaskStatus::CANCELED;
+        }
+      }
+      std::this_thread::sleep_for(100ms);
+    }
+  } catch (nav_core2::PlannerException & e) {
+    RCLCPP_INFO(this->get_logger(), e.what());
+    return TaskStatus::FAILED;
+  }
+
+  nav2_tasks::FollowPathResult result;
+  setResult(result);
+
+  return TaskStatus::SUCCEEDED;
+}
+
+void DwbController::publishVelocity(const nav_2d_msgs::msg::Twist2DStamped & velocity)
+{
+  auto cmd_vel = nav_2d_utils::twist2Dto3D(velocity.velocity);
+  vel_pub_->publish(cmd_vel);
```

Comment:
The publish of the velocities could be via a robot method, where the robot class has the publisher (future TODO).